pkg/network/http/ebpf_gotls.go

```diff
@@ -158,77 +252,122 @@ func supportedArch(arch string) bool {
 	if err != nil {
-		log.Errorf("Could not open file %q due to %s", binPath, err)
+		log.Debugf(" could not read binary path for pid %d: %s", pid, err)
 		return
 	}
-	defer f.Close()
-	elfFile, err := elf.NewFile(f)
-	if err != nil {
-		log.Errorf("File %q could not be parsed as elf: %s", binPath, err)
+
+	var stat syscall.Stat_t
+	if err = syscall.Stat(binPath, &stat); err != nil {
+		log.Errorf("could not stat binary path %s: %s", binPath, err)
 		return
 	}
 
-	result, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
-	if err != nil {
-		log.Errorf("Failed inspecting binary %q: %s", binPath, err)
+	hookedBin := p.analysisCache.Get(stat.Ino)
+	if hookedBin == nil {
+		f, err := os.Open(binPath)
+		if err != nil {
+			log.Errorf("could not open file %s, %s", binPath, err)
+			return
+		}
+		defer f.Close()
+
+		elfFile, err := elf.NewFile(f)
+		if err != nil {
+			log.Errorf("file %s could not be parsed as an ELF file: %s", binPath, err)
+			return
+		}
+
+		inspectionResult, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, structFieldsLookupFunctions)
+		if err != nil {
+			if !errors.Is(err, binversion.ErrNotGoExe) {
+				log.Errorf("error reading exe: %s", err)
+			}
+			return
+		}
+
+		if err = p.addInspectionResultToMap(inspectionResult, stat.Ino); err != nil {
+			log.Error(err)
+			return
+		}
+
+		probeIDs, err := p.attachHooks(inspectionResult, binPath)
+		if err != nil {
+			log.Errorf("error while attaching hooks: %s", err)
+			p.removeInspectionResultFromMap(stat.Ino)
+			return
+		}
+		log.Debugf("attached hooks on %s (%d)", binPath, stat.Ino)
+		hookedBin = p.analysisCache.Set(stat.Ino, &hookedBinary{probeIDs, make(runningProcessesSet)}, ttlcache.NoTTL)
+	} else {
+		log.Debugf("resetting TTL on %s (%d)", binPath, stat.Ino)
+		hookedBin = p.analysisCache.Set(stat.Ino, hookedBin.Value(), ttlcache.NoTTL)
+	}
+
+	hookedBin.Value().running_processes[pid] = struct{}{}
+	p.pidToIno[pid] = stat.Ino
+}
+
+func (p *GoTLSProgram) handleProcessStop(pid pid) {
+	ino, ok := p.pidToIno[pid]
+	if !ok {
 		return
 	}
+	delete(p.pidToIno, pid)
 
-	// result and bin path are being passed as parameters as a preparation for the future when we will have a process
-	// watcher, so we will run on more than one binary in one goTLSProgram.
-	if err := p.addInspectionResultToMap(result, binPath); err != nil {
-		log.Errorf("error in adding inspection result to map: %s", err)
+	hookedBin := p.analysisCache.Get(ino)
+	if hookedBin == nil {
+		log.Error("could not retrieve binary entry from cache")
 		return
 	}
 
-	if err := p.attachHooks(result, binPath); err != nil {
-		log.Errorf("error while attaching hooks: %s", err)
-		p.detachHooks()
+	delete(hookedBin.Value().running_processes, pid)
+	if len(hookedBin.Value().running_processes) == 0 {
+		log.Debugf("no processes left for ino %d", ino)
+		p.analysisCache.Set(ino, hookedBin.Value(), ttlcache.DefaultTTL)
 	}
```

Comment:
so why don't we just delete it?

Reply:
Part of the answer is in the strategy I have written here for the cache: https://github.com/DataDog/datadog-agent/pull/14080#discussion_r1014291380

To add more details on top of it:
I realise here that the name I have chosen for the cache is no longer appropriate. It is not a cache of the analysis results (that we actually do not need anymore after the hooks are attached, so I do not keep them to save a bit on memory usage), but rather a structure keeping track of currently hooked binaries. I'll change it right away!

What we wanted in the first place was to limit the memory usage, and eliminate memory leaks. There was one leak that I fixed early on, another one you highlighted due to the fact we kept binary analysis results forever, and a slightly hidden one with the eBPF map where we store offsets without ever cleaning up stale entries. In the middle of all this, I realised that uprobes are not linked to a specific process, but rather to a specific binary. This means we actually do not need to keep the result of binary analysis around, once the probes are set up and the offsets placed in the eBPF map. This also means we need to keep track of which binaries we hooked, or else we would be trying to attach the same uprobes multiple times, which is not possible and will get ebpf-manager to - rightfully - yell at us. This what the "cache" is used for.

Now, even though we may not encounter many different Go binaries that uses TLS at once, we might not want to keep those entries forever. Since those entries allows us to know whether or not to place the hooks on a specific binary and add offsets in the eBPF map, we can only remove them if there are no running processes associated to that binary. This is why we can't just use a simple LRU cache or TTL, and that I made it so entries with still running processes do not have a TTL and never expire. This also means we must remove the hooks when deleting an entry, and that limiting the size of the cache is directly limiting the number of Go binaries we can hook.

Now, this block of code in `handleProcessStop`, deals with when the number of running processes reaches zero. We have two solutions here:
- Delete the entry here (and so remove the hooks aggressively). It would remove the need for a TTL cache entirely, and we could use a simple Go map instead. It has a flaw though: in the case of a burst of short-lived processes from the same binary running sequentially, or a service written that crashes for some reason, that would mean redoing the whole binary analysis and attaching the hooks each time a new process is spawned. We would have a performance hit in this case, because binary analysis and hooks setup is, relatively speaking, a lengthier process than the second solution...
- Setup a TTL, and wait a bit to see if there are no new processes started from that binary. This is what I do here. In the aforementioned cases, each time a new process is started, we only disable the TTL, and add the PID to a couple of maps to keep track of them. No analysis and no hooks to attach again.

Reply:
If we want to keep track on the binaries that we hook, we don't need that cache, but `map[uint64]struct{}` can suffice
if we want to cache the inspection results, then there is a problem of "memory leak"

let's talk by heart